src/test/java/com/cloudbees/plugins/credentials/CredentialsParametersActionTest.java

```diff
@@ -0,0 +1,120 @@
+import hudson.model.ParametersAction;
+import hudson.model.ParametersDefinitionProperty;
+import hudson.model.User;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+public class CredentialsParametersActionTest {
+
+    @ClassRule public static JenkinsRule j = new JenkinsRule();
+
+    @BeforeClass
+    public static void setUpClass() throws IOException {
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        CredentialsProvider.lookupStores(j.jenkins).iterator().next()
+                .addCredentials(Domain.global(), new UsernamePasswordCredentialsImpl(CredentialsScope.GLOBAL, "cred-id", "global credential", "root", "correct horse battery staple"));
+        final User alpha = User.getOrCreateByIdOrFullName("alpha");
+        try (ACLContext ignored = ACL.as(alpha)) {
+            CredentialsProvider.lookupStores(alpha).iterator().next()
+                    .addCredentials(Domain.global(), new UsernamePasswordCredentialsImpl(CredentialsScope.USER, "alpha-cred-id", "user credentials", "root", "hello world"));
+        }
+    }
+
+    @Test
+    public void forRunReturnsNullWhenRunHasNoParameters() throws Exception {
+        final FreeStyleProject project = j.createFreeStyleProject();
+        final FreeStyleBuild build = j.assertBuildStatusSuccess(project.scheduleBuild2(0));
+        assertNull(CredentialsParametersAction.forRun(build));
+    }
+
+    @Test
+    public void forRunCopiesParametersWhenRunIsParameterized() throws Exception {
+        final FreeStyleProject project = j.createFreeStyleProject();
+        project.addProperty(new ParametersDefinitionProperty(new CredentialsParameterDefinition("cred", "", null, IdCredentials.class.getName(), true)));
+        final FreeStyleBuild build = j.assertBuildStatusSuccess(project.scheduleBuild2(0, new Cause.UserIdCause("alpha"), forParameter("cred-id")));
+        final CredentialsParametersAction action = CredentialsParametersAction.forRun(build);
+        assertNotNull(action);
+        final CredentialsParametersAction.AuthenticatedCredentials creds = action.findCredentialsByParameterName("cred");
+        assertNotNull(creds);
+        assertEquals("alpha", creds.getUserId());
+        assertEquals("cred-id", creds.getParameterValue().getValue());
+    }
+
+    @Test
+    public void forRunUsesUserIdCauseForInitialParametersOwnership() throws Exception {
+        final FreeStyleProject project = j.createFreeStyleProject();
+        project.addProperty(new ParametersDefinitionProperty(new CredentialsParameterDefinition("cred", "", null, IdCredentials.class.getName(), true)));
+
+        // no cause means no user id
+        {
```

Comment:
@jeffret-b are you ok with this construct? I remember to have some comments from you about using methods instead of blocks in test.